File: 01_GOF23/DP_examples/C/02_Structural/Facade/facade.c
#include "facade.h"
#include <stdlib.h>

// 构建形状
IShape *MakeShape(ShapeType shape_type, int *args, int arg_count)
{
    switch (shape_type)
    {
    case CIRCLE:
        if (arg_count >= 1)
            return (IShape *)Circle_create(args[0]);
    case RECTANGLE:
        if (arg_count >= 2)
            return (IShape *)Rectangle_create(args[0], args[1]);
    case SQUARE:
        if (arg_count >= 1)
            return (IShape *)Square_create(args[0]);
    default:
        return NULL;
    }
}

File: 01_GOF23/DP_examples/Rust/02_Structural/Facade/facade.rs
use crate::component::{Circle, IShape, Rectangle, Square};

// 形状类型枚举
#[derive(Debug, Clone, Copy)]
pub enum ShapeType {
    Circle,
    Rectangle,
    Square,
}

pub fn build_shape(shape_type: ShapeType, args: &[i32]) -> Option<Box<dyn IShape>> {
    match shape_type {
        ShapeType::Circle => {
            if args.len() >= 1 {
                Some(Box::new(Circle::new(args[0])))
            } else {
                None
            }
        }
        ShapeType::Rectangle => {
            if args.len() >= 2 {
                Some(Box::new(Rectangle::new(args[0], args[1])))
            } else {
                None
            }
        }
        ShapeType::Square => {
            if args.len() >= 1 {
                Some(Box::new(Square::new(args[0])))
            } else {
                None
            }
        }
    }
}
